packages/esm-appointments-app/src/appointments/scheduled/scheduled-appointments.component.tsx

```diff
@@ -16,101 +23,165 @@ interface ScheduledAppointmentsProps {
-                  status: extension.config.status,
-                  title: extension.config.title,
-                }}
-              />
-            </div>
-          )
-        }
+        {(extension) => {
+          return (
+            <ExtensionWrapper
+              extension={extension}
+              currentTab={currentTab}
+              appointmentServiceType={appointmentServiceType}
+              date={date}
+              dateType={dateType}
+              showExtensionTab={showExtension}
+              hideExtensionTab={hideExtension}
+            />
+          );
+        }}
       </ExtensionSlot>
     </>
   );
 };
 
+function useAllowedExtensions() {
+  const [allowedExtensions, setAllowedExtensions] = useState<Readonly<Record<string, boolean>>>({});
+
+  const set = useCallback(
+    (value: boolean, key: string) => {
+      setAllowedExtensions({ ...allowedExtensions, [key]: value });
+    },
+    [allowedExtensions, setAllowedExtensions],
+  );
+
+  return {
+    allowedExtensions,
+    showExtension: set.bind(null, true) as (extension: string) => void,
+    hideExtension: set.bind(null, false) as (extension: string) => void,
+  };
+}
+
+function ExtensionWrapper({
+  extension,
+  currentTab,
+  appointmentServiceType,
+  date,
+  dateType,
+  showExtensionTab,
+  hideExtensionTab,
+}: {
+  extension: ConnectedExtension;
+  currentTab: string;
+  appointmentServiceType: string;
+  date: string;
+  dateType: DateType;
+  showExtensionTab: (extension: string) => void;
+  hideExtensionTab: (extension: string) => void;
+}) {
+  const currentConfig = useRef(null);
+
+  // This use effect hook controls whether the tab for this extension should render
+  useEffect(() => {
+    if (
+      currentConfig.current === null ||
+      (currentConfig.current !== null && !shallowEqual(currentConfig.current, extension.config))
+    ) {
+      currentConfig.current = extension.config;
+      shouldDisplayExtensionTab(extension?.config, dateType)
+        ? showExtensionTab(extension.name)
+        : hideExtensionTab(extension.name);
+    }
+  }, [extension, date, dateType, currentTab, showExtensionTab, hideExtensionTab]);
+
+  return (
+    <div
+      key={extension.name}
+      className={styles.container}
+      style={{ display: currentTab === extension.name ? 'block' : 'none' }}>
+      <Extension
+        state={{
+          date,
+          appointmentServiceType,
+          status: extension.config?.status,
+          title: extension.config?.title,
+        }}
+      />
+    </div>
+  );
+}
+
+function shouldDisplayExtensionTab(config: ConfigObject | undefined, dateType: DateType): boolean {
+  if (!config) {
+    return false;
+  }
+
+  switch (dateType) {
+    case 'futureDate':
+      return config.showForFutureDate ?? false;
+    case 'pastDate':
+      return config.showForPastDate ?? false;
+    case 'today':
+      return config.showForToday ?? false;
+  }
+}
+
+function shallowEqual(objA: object, objB: object) {
```

Comment:
I'm going to add this to `esm-utils`, but it's here for now. This just checks if two objects are "shallowly equal" to one another. This is basically the algorithm React uses internally to determine whether to re-render a component based on the component's props or state.